cpp/src/arrow/result.h

```diff
@@ -371,7 +371,7 @@ class [[nodiscard]] Result : public util::EqualityComparable<Result<T>> {
     if (ok()) {
       return MoveValueUnsafe();
     }
-    return generate_alternative();
+    return std::forward<G>(generate_alternative)();
```

Comment:
Note that from the comment, `generate_alternative` is supposed to be a factory not an actual `G` or parameters that can be forwarded to `G`'s constructor. This is a way to avoid evaluating the parameter when `Result` already contains something. 

Functional Programming languages make this kind of lazy parameter (call-by-name) very easy to declare. In scalar this would be `alternative: => T` instead of `alternative: T`.

https://www.scala-lang.org/api/current/scala/Option.html#getOrElse[B%3E:A](default:=%3EB):B

Reply:
@PragmaTwice a way to fix the linked issue would be to ensure somehow that `G` is a lambda.

Reply:
An `enable_if` using `is_invocable_r` would restrict the calls only to generators `G` that return `T`.

[1] https://en.cppreference.com/w/cpp/types/is_invocable